src/main/kotlin/com/cognifide/gradle/aem/common/instance/InstanceIMSClient.kt

```diff
@@ -0,0 +1,161 @@
+        clientSecret = integration.getJSONObject("technicalAccount").getString("clientSecret")
+        imsHost = integration.getString("imsEndpoint")
+        imsExchangeEndpoint = "https://$imsHost/ims/exchange/jwt"
+        metaScopes = integration.getString("metascopes").split(",")
+    }
+
+    private fun getJWTToken(): String {
+        var privateKeyContent: ByteArray
+        PEMParser(StringReader(privateKey)).use { pemParser ->
+            val keyPair: KeyPair = JcaPEMKeyConverter().getKeyPair(pemParser.readObject() as PEMKeyPair)
+            privateKeyContent = keyPair.private.encoded
+        }
+        val keyFactory = KeyFactory.getInstance("RSA")
+        val keySpec: KeySpec = PKCS8EncodedKeySpec(privateKeyContent)
+        val rsaPrivateKey = keyFactory.generatePrivate(keySpec) as RSAPrivateKey
+
+        val jwtClaims: MutableMap<String, Any> = HashMap()
+        jwtClaims["iss"] = orgId
+        jwtClaims["sub"] = technicalAccountId
+        jwtClaims["exp"] = expirationTime
+        jwtClaims["aud"] = "https://$imsHost/c/$clientId"
+        for (scope in metaScopes) {
+            jwtClaims["https://$imsHost/s/$scope"] = true
+        }
+
+        return Jwts.builder()
+            .setClaims(jwtClaims)
+            .signWith(SignatureAlgorithm.RS256, rsaPrivateKey)
+            .compact()
+    }
+
+    private fun getAccessToken(jwtToken: String): String? {
+        val connection = URL(imsExchangeEndpoint).openConnection() as HttpsURLConnection
+        connection.requestMethod = HttpMethods.POST
+        val urlParameters = "client_id=$clientId&client_secret=$clientSecret&jwt_token=$jwtToken"
+
+        // Send post request
+        connection.doOutput = true
+        val writer = DataOutputStream(connection.outputStream)
+        writer.writeBytes(urlParameters)
+        writer.flush()
+        writer.close()
+
+        var responseError = false
+
+        val inputStream: InputStream
+        if (connection.responseCode < HttpsURLConnection.HTTP_BAD_REQUEST) {
+            inputStream = connection.inputStream
+        } else {
+            inputStream = connection.errorStream
+            responseError = true
+        }
+
+        val response = inputStream.bufferedReader().use { it.readText() }
+
+        if (responseError) {
+            println(response)
+            return null
+        }
+
+        return JSONObject(response).getString("access_token")
+    }
+
+    companion object {
+
+        @Suppress("TooGenericExceptionCaught")
+        fun generateBearerTokenOrNull(aem: AemExtension): String? {
+            val client = InstanceIMSClient(aem)
+            with(client) {
+                if (keyPath.isPresent) {
+                    try {
+                        readProperties()
+                        val jwtToken = getJWTToken()
+                        return getAccessToken(jwtToken)
+                    } catch (e: Exception) {
+                        println("Couldn't generate the access token")
```

Comment:
Because when I throw an exception, I end up with failed task when I run "props" because this code runs every time I run a task which makes the user not able to launch props again to change something (for example choosing a different token file) so I didn't know how to handle it other than just printing to console.
What do you suggest I do?

Reply:
not sure how to do that, any suggestions on where to look?